datashape/predicates.py

```diff
@@ -104,7 +109,6 @@ def _dimensions(ds):
         return 1 + _dimensions(ds.subshape[0])
     if isscalar(ds):
         return 0
-    raise NotImplementedError('Can not compute dimensions for %s' % ds)
```

Comment:
i forgot to add a test that exercises this code and i saw that it wasn't being hit. it's definitely possible to get here and we don't necessarily expect one of the branches to return. a function prototype is an example of something that doesn't have dimensions